main.go

```diff
@@ -95,114 +94,173 @@ func getEtcdConfig() (etcd.Config, error) {
+
+	globalConfig, err := s.getGlobalConfig()
+	if err != nil {
+		log.Fatal(err)
 	}
-	return uint32(asn), nil
+
+	if err := s.bgpServer.Start(globalConfig); err != nil {
+		log.Fatal(err)
+	}
+
+	// monitor routes from other BGP peers and update FIB
+	s.t.Go(s.monitorPath)
+	// watch prefix assigned and announce to other BGP peers
+	s.t.Go(s.watchPrefix)
+	// watch BGP configuration
+	s.t.Go(s.watchBGPConfig)
+
+	<-s.t.Dying()
+	log.Fatal(s.t.Err())
+
+}
+
+func (s *Server) getGlobalASN() (numorstring.ASNumber, error) {
+	return s.cli.Config().GetGlobalASNumber()
 }
 
-func getPeerASN(api etcd.KeysAPI, host string) (uint32, error) {
-	res, err := api.Get(context.Background(), fmt.Sprintf("%s/host/%s/as_num", CALICO_BGP, host), nil)
-	if errorButKeyNotFound(err) != nil {
+func (s *Server) getPeerASN(host string) (numorstring.ASNumber, error) {
+	node, err := s.cli.Nodes().Get(calicoapi.NodeMetadata{Name: host})
+	if err != nil {
 		return 0, err
 	}
-	if res != nil && res.Node != nil {
-		v, err := strconv.ParseUint(res.Node.Value, 10, 32)
-		if err != nil {
-			return 0, err
-		}
-		return uint32(v), nil
+	if node.Spec.BGP == nil {
+		return 0, fmt.Errorf("host %s is running in policy-only mode")
+	}
+	asn := node.Spec.BGP.ASNumber
+	if asn == nil {
+		return s.getGlobalASN()
 	}
-	return getGlobalASN(api)
+	return *asn, nil
 
 }
 
-func getGlobalConfig(api etcd.KeysAPI) (*bgpconfig.Global, error) {
-	asn, err := getGlobalASN(api)
+func (s *Server) getGlobalConfig() (*bgpconfig.Global, error) {
+	asn, err := s.getGlobalASN()
 	if err != nil {
 		return nil, err
 	}
 	return &bgpconfig.Global{
 		Config: bgpconfig.GlobalConfig{
-			As:       asn,
-			RouterId: os.Getenv(IP),
+			As:       uint32(asn),
+			RouterId: s.ipv4.String(),
 		},
 	}, nil
 }
 
-func isMeshMode(api etcd.KeysAPI) (bool, error) {
-	res, err := api.Get(context.Background(), fmt.Sprintf("%s/global/node_mesh", CALICO_BGP), nil)
-	if err != nil {
-		return false, err
-	}
-	m := &struct {
-		Enabled bool `json:"enabled"`
-	}{}
-	if err := json.Unmarshal([]byte(res.Node.Value), m); err != nil {
-		return false, err
-	}
-	return m.Enabled, nil
+func (s *Server) isMeshMode() (bool, error) {
+	return s.cli.Config().GetNodeToNodeMesh()
 }
 
-func getMeshNeighborConfigs(api etcd.KeysAPI) ([]*bgpconfig.Neighbor, error) {
-	globalASN, err := getGlobalASN(api)
+func (s *Server) getMeshNeighborConfigs() ([]*bgpconfig.Neighbor, error) {
+	globalASN, err := s.getGlobalASN()
 	if err != nil {
 		return nil, err
 	}
-	res, err := api.Get(context.Background(), fmt.Sprintf("%s/host", CALICO_BGP), &etcd.GetOptions{Recursive: true})
+	nodes, err := s.cli.Nodes().List(calicoapi.NodeMetadata{})
 	if err != nil {
 		return nil, err
 	}
-	ns := make([]*bgpconfig.Neighbor, 0, len(res.Node.Nodes))
-	for _, node := range res.Node.Nodes {
-		var v4, v6 string
+	ns := make([]*bgpconfig.Neighbor, 0, len(nodes.Items))
+	for _, node := range nodes.Items {
+		if node.Metadata.Name == os.Getenv(NODENAME) {
```

Comment:
I wonder if this is the best way to check if it's the node itself. Thinking out loud here since we use to have the ENV var called "HOSTNAME" before, not sure to which extent we keep that backward compatibility. @robbrockbank will `os.Getenv(NODENAME)` always result in what is intended here?